MuEnvironment/GitDependency.py

```diff
@@ -0,0 +1,104 @@
+# IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
+# INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+# DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+# LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE
+# OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
+# ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+##
+import os
+import logging
+from MuEnvironment.ExternalDependency import ExternalDependency
+from MuEnvironment import RepoResolver
+from MuEnvironment.MuGit import Repo
+from MuEnvironment import VersionAggregator
+
+
+class GitDependency(ExternalDependency):
+    '''
+    ext_dep fields:
+    - source:  url for git clone
+    - version: commit from git repo
+    '''
+
+    TypeString = "git"
+
+    def __init__(self, descriptor):
+        super().__init__(descriptor)
+        self.repo_url = self.source
+        self.commit = self.version
+        self._local_repo_root_path = os.path.join(os.path.abspath(self.contents_dir), self.name)
+        self.logger = logging.getLogger("git-dependency")
+
+        # valid_attributes = ["Path", "Url", "Branch", "Commit", "ReferencePath", "Full"]
+        self._repo_resolver_dep_obj = {"Path": self.name, "Url": self.repo_url, "Commit": self.commit}
+
+    def fetch(self):
+
+        # def resolve(file_system_path, dependency, force=False, ignore=False, update_ok=False):
+        RepoResolver.resolve(self._local_repo_root_path, self._repo_resolver_dep_obj, update_ok=True)
+
+        # Add a file to track the state of the dependency.
+        self.update_state_file()
+
+    def clean(self):
+        self.logger.debug("Cleaning git dependency directory for '%s'..." % self.name)
+
+        if os.path.isdir(self._local_repo_root_path):
+            # Clean up git dependency specific stuff
+            RepoResolver.clear_folder(self.contents_dir)
+
+        # Let super class clean up common dependency stuff
+        super().clean()
+
+    # override verify due to different scheme with git
+    def verify(self, logversion=True):
+        result = True
+
+        if not os.path.isdir(self._local_repo_root_path):
+            self.logger.error("no dir for Git Dependency")
+            result = False
+
+        if result and len(os.listdir(self._local_repo_root_path)) == 0:
+            self.logger.error("no files in Git Dependency")
+            result = False
+
+        if result:
+            # valid repo folder
+            r = Repo(self._local_repo_root_path)
+            if(not r.initalized):
+                self.logger.error("Git Dependency: Not Initialized")
+                result = False
+            elif(r.dirty):
+                self.logger.error("Git Dependency: dirty")
+                result = False
+
+            if(r.head.commit != self.version):
```

Comment:
So we only support tracking a specific commit, do we plan on tracking a specific branch?

Reply:
Ext_dep fit better with just a commit as version.  I think it should be put on the list to be added later